### PR TITLE
Replace RowReceiver with BatchIterator/BatchConsumer

### PR DESCRIPTION
**Pull request by mfussenegger**
Migrates from a push based execution pipeline to pull based.

### CODE REVIEW

File: dex/src/main/java/io/crate/data/BatchConsumer.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+ *     An exception to this is if the throwable in the {@link #accept(BatchIterator, Throwable)} call isn't null.
+ *     In that case the Consumer is not required to call close, as the iterator may even be null.
+ * </p>
+ * <p>
+ *     Multiple calls to {@link #accept(BatchIterator, Throwable)} are not allowed.
+ * </p>
+ */
+
+/*
+@startuml
```
Comment (seut) [inline dex/src/main/java/io/crate/data/BatchConsumer.java]: pls remove uml as discussed

File: dex/src/main/java/io/crate/data/Columns.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+     * @return the input at the specified position
+     * @throws IndexOutOfBoundsException if the index is out of range
+     *                                   (<tt>index &lt; 0 || index &gt;= size()</tt>)
+     */
+    Input<?> get(int index);
+
+    /**
+     * Returns the number of columns.
+     *
+     * @return the number of columns
```
Comment (seut) [inline dex/src/main/java/io/crate/data/Columns.java]: isn't this a bit over-documenting ? ;)

File: dex/src/main/java/io/crate/data/Columns.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    /**
+     * Creates a new columns object which wraps the given list of inputs. Modifications to the passed in list must be
+     * prevented, since this would break the immutability contract of this interface.
+     *
+     * @param listOfInputs the list of inputs to be wrapped
+     * @return a new columns object
+     */
+    static Columns wrap(List<? extends Input<?>> listOfInputs) {
```
Comment (seut) [inline dex/src/main/java/io/crate/data/Columns.java]: hm calling it just `inputs` is fine imho, list type is already known due signature

File: build.gradle [hunk 4]
```diff
@@ -8,7 +8,8 @@ buildscript {
 }
 
 plugins {
-    id 'me.champeau.gradle.jmh' version '0.3.0'
+    id 'com.github.johnrengelman.shadow' version '1.2.4'
```
Comment (matriv) [inline build.gradle]: can we move these versions to our versions properties file?

**Pull request merged**